temporian/beam/io.py

```diff
@@ -0,0 +1,299 @@
+
+    This format is similar to output of the official beam IO connectors:
+    https://beam.apache.org/documentation/io/connectors/
+
+    CSV values are always string, so the output of `read_csv_raw` is always
+    a dictionary of string to string. Use `to_event_set` (or better, use
+    `read_csv` instead of `read_csv_raw`) to cast values to the expected
+    pipeline input dtype.
+
+    Args:
+        pipe: A begin Beam pipe.
+        file_pattern: Path or path matching expression compatible with
+            `MatchFiles`.
+
+    Returns:
+        A PCollection of dictionary of key:value.
+    """
+
+    return (
+        pipe
+        | "List files" >> MatchFiles(file_pattern)
+        | "Parse file" >> beam.FlatMap(_parse_csv_file)
+    )
+
+
+def _cast_feature_value(value: Any, dtype: DType) -> np.generic:
+    """Convert a user feature value to the internal representation."""
+
+    np_type = tp_dtype_to_np_dtype(dtype)
+    return np_type(value)
+
+
+def _cast_index_value(value: Any, dtype: DType) -> BeamIndex:
+    """Convert a user index value to the internal representation."""
+
+    return _cast_feature_value(value, dtype).item()
+
+
+def _structure_fm(
+    row: Dict[str, str], schema: Schema, timestamp_key: str
+) -> StructuredRow:
+    """Transforms a dict of key:value to a StructuredRow."""
+
+    index_values = [
+        _cast_index_value(row[index.name], index.dtype)
+        for index in schema.indexes
+    ]
+    feature_values = [
+        _cast_feature_value(row[feature.name], feature.dtype)
+        for feature in schema.features
+    ]
+    timestamp = np.float64(row[timestamp_key])
+    return tuple(index_values), (timestamp, tuple(feature_values))
+
+
+class _MergeTimestampsDP(beam.DoFn):
+    """Aggregates StructuredRows into BeamEventSet."""
+
+    def __init__(self, num_features: int):
+        self._num_features = num_features
+
+    def process(
+        self,
+        item: Tuple[
+            Tuple[BeamIndex, ...],
+            Iterable[Tuple[Timestamps, Tuple[np.generic, ...]]],
+        ],
+    ) -> Iterable[BeamEventSet]:
+        index, feat_and_ts = item
+        timestamps = np.array([v[0] for v in feat_and_ts], dtype=np.float64)
+        for feature_idx in range(self._num_features):
+            values = np.array([v[1][feature_idx] for v in feat_and_ts])
+            yield index + (feature_idx,), (timestamps, values)
+
+
+@beam.ptransform_fn
+def to_event_set(
+    pipe: beam.PCollection[Dict[str, Any]],
+    schema: Schema,
+    timestamp_key: str = "timestamp",
+) -> beam.PCollection[BeamEventSet]:
+    """Converts a PCollection of key:value to a Beam event-set.
+
+    This method is compatible with the output of `read_csv_raw` and the
+    Official Beam IO connectors.
+
+    When importing data from csv files, use `read_csv` to convert csv files
+    directly into event sets.
+
+    Unlike Temporian in-process event set import method (`tp.event_set`), this
```

Comment:
use cross-refs for navigable docs (in this case, s/`tp.event_set`/[`tp.event_set`][temporian.event_set])

Reply:
*Done